[Cutlass 3.2.2 submodule upgrade] Adapt Inductor cutlass backend to Cutlass 3.2.2

kadeng opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112861
* #110890
* __->__ #112762

The inductor cutlass backend was written against Cutlass version 3.1.x,
there are some incompatible changes in Cutlass 3.2.2 which the
Inductor cutlass backend needs to adapt to.

Test plan:

If third_party/cutlass is upgraded to Cutlass tag v3.2.2,
several tests within test/inductor/test_max_autotune.py start to
fail. With this diff applied, they pass again.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @muchulee8 @aakhundov @ColinPeppler

Differential Revision: [D50986555](https://our.internmc.facebook.com/intern/diff/D50986555)

kadeng committed:
[Cutlass 3.2.2 upgrade] Adapt Inductor cutlass backend to Cutlass 3.2.2

The inductor cutlass backend was written against Cutlass version 3.1.x,
there are some incompatible changes in Cutlass 3.2.2 which the
Inductor cutlass backend needs to adapt to.

Test plan:

If third_party/cutlass is upgraded to Cutlass tag v3.2.2,
several tests within test/inductor/test_max_autotune.py start to
fail. With thid diff, they pass again.

[ghstack-poisoned]

aakhundov commented:
Thanks @kadeng! Do we need to enable `ciflow/trunk`, too?

kadeng commented:
Currently a Windows build failure is blocking this.  A Cutlass Issue has been created here: https://github.com/NVIDIA/cutlass/issues/1171

kadeng committed:
Attempted hotfix of Cutlass 3.2.2 on "[Cutlass 3.2.2 upgrade] Adapt Inductor cutlass backend to Cutlass 3.2.2"


The inductor cutlass backend was written against Cutlass version 3.1.x,
there are some incompatible changes in Cutlass 3.2.2 which the
Inductor cutlass backend needs to adapt to.

Test plan:

If third_party/cutlass is upgraded to Cutlass tag v3.2.2,
several tests within test/inductor/test_max_autotune.py start to
fail. With this diff applied, they pass again.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Attempted hotfix (attempt 2) of Cutlass 3.2.2 on "[Cutlass 3.2.2 upgrade] Adapt Inductor cutlass backend to Cutlass 3.2.2"


The inductor cutlass backend was written against Cutlass version 3.1.x,
there are some incompatible changes in Cutlass 3.2.2 which the
Inductor cutlass backend needs to adapt to.

Test plan:

If third_party/cutlass is upgraded to Cutlass tag v3.2.2,
several tests within test/inductor/test_max_autotune.py start to
fail. With this diff applied, they pass again.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Fixing MSVC __cplusplus behavior as recommended for Cutlass on "[Cutlass 3.2.2 upgrade] Adapt Inductor cutlass backend to Cutlass 3.2.2"


The inductor cutlass backend was written against Cutlass version 3.1.x,
there are some incompatible changes in Cutlass 3.2.2 which the
Inductor cutlass backend needs to adapt to.

Test plan:

If third_party/cutlass is upgraded to Cutlass tag v3.2.2,
several tests within test/inductor/test_max_autotune.py start to
fail. With this diff applied, they pass again.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Cutlass 3.2.2 upgrade] Adapt Inductor cutlass backend to Cutlass 3.2.2"


The inductor cutlass backend was written against Cutlass version 3.1.x,
there are some incompatible changes in Cutlass 3.2.2 which the
Inductor cutlass backend needs to adapt to.

Test plan:

If third_party/cutlass is upgraded to Cutlass tag v3.2.2,
several tests within test/inductor/test_max_autotune.py start to
fail. With this diff applied, they pass again.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Cutlass 3.2.2 upgrade] Adapt Inductor cutlass backend to Cutlass 3.2.2"


The inductor cutlass backend was written against Cutlass version 3.1.x,
there are some incompatible changes in Cutlass 3.2.2 which the
Inductor cutlass backend needs to adapt to.

Test plan:

If third_party/cutlass is upgraded to Cutlass tag v3.2.2,
several tests within test/inductor/test_max_autotune.py start to
fail. With this diff applied, they pass again.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng commented on CMakeLists.txt:
```diff
@@ -479,6 +479,17 @@ if(LINUX)
 endif()
 
 if(MSVC)
+  # MSVC by default does not apply the correct __cplusplus version as specified by the C++ standard
```
Resolving Windows build issue, see https://github.com/NVIDIA/cutlass/issues/1171#issuecomment-1792411693

drisspg replied:
If this does fix the problem I am not sure if we want to set this global, @malfet will likely have thoughts on this.

That being said it is possible to set flags on specific targets, see:
https://github.com/pytorch/pytorch/blob/d084a024aede3b2420c5977dd473988c63e9b9f6/caffe2/CMakeLists.txt#L1820-L1832

kadeng replied:
As mentioned in the chat, I think we want to set this globally, since it restores MSVC to C++-Standard compliant behavior. 

See 
 * [C++ Standard](https://en.cppreference.com/w/cpp/preprocessor/replace#Predefined_macros)
 * [MSVC Docs](https://learn.microsoft.com/en-us/cpp/build/reference/zc-cplusplus?view=msvc-170) which pretty clearly states that the default behavior of MSVC is a hack.

ipiszy reviewed (approved):
Thanks @kadeng !

kadeng committed:
Update on "[Cutlass 3.2.2 upgrade] Adapt Inductor cutlass backend to Cutlass 3.2.2"


The inductor cutlass backend was written against Cutlass version 3.1.x,
there are some incompatible changes in Cutlass 3.2.2 which the
Inductor cutlass backend needs to adapt to.

Test plan:

If third_party/cutlass is upgraded to Cutlass tag v3.2.2,
several tests within test/inductor/test_max_autotune.py start to
fail. With this diff applied, they pass again.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Cutlass 3.2.2 upgrade] Adapt Inductor cutlass backend to Cutlass 3.2.2"


The inductor cutlass backend was written against Cutlass version 3.1.x,
there are some incompatible changes in Cutlass 3.2.2 which the
Inductor cutlass backend needs to adapt to.

Test plan:

If third_party/cutlass is upgraded to Cutlass tag v3.2.2,
several tests within test/inductor/test_max_autotune.py start to
fail. With this diff applied, they pass again.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng commented:
@kadeng has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50986555).

kadeng commented:
@pytorchbot merge

kadeng commented:
@pytorchbot merge

kadeng commented:
Just for the record, this PR passed all tests. The test failures seem to be due to a temporary github outage.

kadeng commented:
Triggered re-run of all checks...

kadeng committed:
Triggering new checks after github outage on "[Cutlass 3.2.2 submodule upgrade] Adapt Inductor cutlass backend to Cutlass 3.2.2"


The inductor cutlass backend was written against Cutlass version 3.1.x,
there are some incompatible changes in Cutlass 3.2.2 which the
Inductor cutlass backend needs to adapt to.

Test plan:

If third_party/cutlass is upgraded to Cutlass tag v3.2.2,
several tests within test/inductor/test_max_autotune.py start to
fail. With this diff applied, they pass again.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

Differential Revision: [D50986555](https://our.internmc.facebook.com/intern/diff/D50986555)

[ghstack-poisoned]

kadeng commented:
@kadeng has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50986555).

kadeng commented:
@pytorchbot merge

kadeng committed:
Update on "[Cutlass 3.2.2 submodule upgrade] Adapt Inductor cutlass backend to Cutlass 3.2.2"


The inductor cutlass backend was written against Cutlass version 3.1.x,
there are some incompatible changes in Cutlass 3.2.2 which the
Inductor cutlass backend needs to adapt to.

Test plan:

If third_party/cutlass is upgraded to Cutlass tag v3.2.2,
several tests within test/inductor/test_max_autotune.py start to
fail. With this diff applied, they pass again.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

Differential Revision: [D50986555](https://our.internmc.facebook.com/intern/diff/D50986555)

[ghstack-poisoned]

kadeng commented:
@kadeng has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50986555).

kadeng commented:
@kadeng has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50986555).

kadeng commented:
@kadeng has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50986555).

kadeng commented:
@pytorchbot merge
